docs: add page for starting a custom chain (#1369)
FuelLabs/fuels-ts: apps/docs/src/guide/testing/setting-up-a-custom-chain.md

# Setting up a custom chain

THe `launchNodeAndGetWallets` method lets you launch a local Fuel node with various customizations.

In the code snippet below, we provide a custom chain config file to the `launchNodeAndGetWallets` method. You can use a chain config file to customize things like the chain's consensus parameters or specify some initial states for the chain. Click here to see what a chain config file looks like: [chainConfig.json](https://github.com/FuelLabs/fuels-ts/blob/master/.fuel-core/configs/chainConfig.json)

> <<< @/../../../packages/wallet/src/test-utils/launchNode.test.ts#launchNode-custom-config{ts:line-numbers}

Reviewer: Should this snippet be below the `## Customization Options` heading?

Author: No, this snippet only shows the basic usage and touches upon only one of the customization options there. The `## Customization Options` section covers all the options separately

Reviewer 2: @Dhaiwat10 Apologies, I wasn't clear enough.

My concern is that people might not know what options like `--manual_blocks_enabled` means.

I was wondering if we could add some links to `fuel-core` docs, but I found nothing:
 - https://docs.fuel.network/guides/running-a-node/
 - https://docs.fuel.network/guides/running-a-node/running-a-local-node/

Perhaps we could mention that they can see all the options by running something like:

```console
pnpm fuels core run -h
```

Which would give us this:

<details>
  <summary>show code snippet</summary>
  
```
Run the Fuel client node locally

Usage: fuel-core run [OPTIONS]

Options:
      --ip <IP>
          [env: IP=] [default: 127.0.0.1]
      --port <PORT>
          [env: PORT=] [default: 4000]
      --service-name <SERVICE_NAME>
          Vanity name for node, used in telemetry [env: SERVICE_NAME=] [default: fuel-core]
      --max-database-cache-size <MAX_DATABASE_CACHE_SIZE>
          The maximum database cache size in bytes [env: MAX_DATABASE_CACHE_SIZE=] [default: 1073741824]
      --db-path <DB_PATH>
          [env: DB_PATH=] [default: /Users/anderson/.fuel/db]
      --db-type <DATABASE_TYPE>
          [env: DATABASE_TYPE=] [default: rocks-db] [possible values: in-memory, rocks-db]
      --chain <CHAIN_CONFIG>
          Specify either an alias to a built-in configuration or filepath to a JSON file [env: CHAIN_CONFIG=] [default: local_testnet]
      --manual_blocks_enabled
          Allows GraphQL Endpoints to arbitrarily advanced blocks. Should be used for local development only [env: MANUAL_BLOCKS_ENABLED=]
      --vm-backtrace
          Enable logging of backtraces from vm errors [env: VM_BACKTRACE=]
      --utxo-validation
          Enable full utxo stateful validation disabled by default until downstream consumers stabilize [env: UTXO_VALIDATION=]
      --min-gas-price <MIN_GAS_PRICE>
          The minimum allowed gas price [env: MIN_GAS_PRICE=] [default: 0]
      --consensus-key <CONSENSUS_KEY>
          The signing key used when producing blocks. Setting via the `CONSENSUS_KEY_SECRET` ENV var is preferred [env: CONSENSUS_KEY=]
      --poa-instant <INSTANT>
          Use instant block production mode. Newly submitted txs will immediately trigger the production of the next block. Cannot be combined with other poa flags [env: INSTANT=] [default: true] [possible values: true, false]
      --poa-interval-period <PERIOD>
          Interval trigger option. Produces blocks on a fixed interval regardless of txpool activity. Cannot be combined with other poa flags [env: PERIOD=]
      --dev-keys
          Use a default insecure consensus key for testing purposes. This will not be enabled by default in the future [env: CONSENSUS_DEV_KEY=]
      --coinbase-recipient <COINBASE_RECIPIENT>
          The block's fee recipient public key [env: COINBASE_RECIPIENT=]
      --relayer <ETH_CLIENT>
          Uri address to ethereum client. It can be in format of `http://localhost:8545/` or `ws://localhost:8545/`. If not set relayer will not start [env: ETH_CLIENT=]
      --relayer-v2-listening-contracts <ETH_V2_LISTENING_CONTRACTS>
          Ethereum contract address. Create EthAddress into fuel_types [env: ETH_V2_LISTENING_CONTRACTS=]
      --relayer-da-finalization <DA_FINALIZATION>
          Number of da block after which messages/stakes/validators become finalized [env: DA_FINALIZATION=] [default: 100]
      --relayer-da-deploy-height <DA_DEPLOY_HEIGHT>
          Number of da block that the contract is deployed at [env: DA_DEPLOY_HEIGHT=] [default: 0]
      --relayer-log-page-size <LOG_PAGE_SIZE>
          Number of pages or blocks containing logs that should be downloaded in a single call to the da layer [env: LOG_PAGE_SIZE=] [default: 10000]
      --relayer-min-duration-s <SYNC_MINIMUM_DURATION_SECS>
          The minimum number of seconds that the relayer polling loop will take before running again. If this is too low the DA layer risks being spammed [env: SYNC_MINIMUM_DURATION_SECS=] [default: 5]
      --relayer-eth-sync-call-freq-s <SYNCING_CALL_FREQUENCY_SECS>
          [env: SYNCING_CALL_FREQUENCY_SECS=] [default: 5]
      --relayer-eth-sync-log-freq-s <SYNCING_LOG_FREQUENCY_SECS>
          [env: SYNCING_LOG_FREQUENCY_SECS=] [default: 60]
  -e, --enable-p2p
          Enable P2P. By default, P2P is disabled, even when the binary is compiled with the "p2p" feature flag. Providing `--enable-p2p` will enable the P2P service
      --keypair <KEYPAIR>
          Peering secret key. Supports either a hex encoded secret key inline or a path to bip32 mnemonic encoded secret file [env: KEYPAIR=]
      --network <NETWORK>
          The name of the p2p Network [env: NETWORK=]
      --address <ADDRESS>
          p2p network's IP Address [env: ADDRESS=]
      --public_address <PUBLIC_ADDRESS>
          Optional address of your local node made reachable for other nodes in the network [env: PUBLIC_ADDRESS=]
      --peering_port <PEERING_PORT>
          p2p network's TCP Port [env: PEERING_PORT=] [default: 30333]
      --max_block_size <MAX_BLOCK_SIZE>
          Max Block size [env: MAX_BLOCK_SIZE=] [default: 18874368]
      --bootstrap_nodes <BOOTSTRAP_NODES>
          Addresses of the bootstrap nodes They should contain PeerId within their `Multiaddr` [env: BOOTSTRAP_NODES=]
      --reserved_nodes <RESERVED_NODES>
          Addresses of the reserved nodes They should contain PeerId within their `Multiaddr` [env: RESERVED_NODES=]
      --reserved_nodes_only_mode
          With this set to `true` you create a guarded node that is only ever connected to trusted, reserved nodes [env: RESERVED_NODES_ONLY_MODE=]
      --enable_mdns
          Allow nodes to be discoverable on the local network [env: ENABLE_MDNS=]
      --max_peers_connected <MAX_PEERS_CONNECTED>
          Max number of unique peers connected This number should be at least number of `mesh_n` from `Gossipsub` configuration. The total number of connections will be `(max_peers_connected + reserved_nodes.len()) * max_connections_per_peer` [env: MAX_PEERS_CONNECTED=] [default: 50]
      --max_connections_per_peer <MAX_CONNECTIONS_PER_PEER>
          Max number of connections per single peer The total number of connections will be `(max_peers_connected + reserved_nodes.len()) * max_connections_per_peer` [env: MAX_CONNECTIONS_PER_PEER=] [default: 3]
      --random_walk <RANDOM_WALK>
          Set the delay between random walks for p2p node discovery in seconds. If it's not set the random walk will be disabled. Also if `reserved_nodes_only_mode` is set to `true`, the random walk will be disabled [env: RANDOM_WALK=] [default: 0]
      --allow_private_addresses
          Choose to include private IPv4/IPv6 addresses as discoverable except for the ones stored in `bootstrap_nodes` [env: ALLOW_PRIVATE_ADDRESSES=]
      --connection_idle_timeout <CONNECTION_IDLE_TIMEOUT>
          Choose how long will connection keep alive if idle [env: CONNECTION_IDLE_TIMEOUT=] [default: 120]
      --info_interval <INFO_INTERVAL>
          Choose how often to recieve PeerInfo from other nodes [env: INFO_INTERVAL=] [default: 3]
      --identify_interval <IDENTIFY_INTERVAL>
          Choose the interval at which identification requests are sent to the remote on established connections after the first request [env: IDENTIFY_INTERVAL=] [default: 5]
      --max_mesh_size <MAX_MESH_SIZE>
          Choose max mesh size for gossipsub protocol [env: MAX_MESH_SIZE=] [default: 12]
      --min_mesh_size <MIN_MESH_SIZE>
          Choose min mesh size for gossipsub protocol [env: MIN_MESH_SIZE=] [default: 4]
      --ideal_mesh_size <IDEAL_MESH_SIZE>
          Choose ideal mesh size for gossipsub protocol [env: IDEAL_MESH_SIZE=] [default: 6]
      --history_length <HISTORY_LENGTH>
          Number of heartbeats to keep in the gossipsub `memcache` [env: HISTORY_LENGTH=] [default: 5]
      --history_gossip <HISTORY_GOSSIP>
          Number of past heartbeats to gossip about [env: HISTORY_GOSSIP=] [default: 3]
      --gossip_heartbeat_interval <GOSSIP_HEARTBEAT_INTERVAL>
          Time between each gossipsub heartbeat [env: GOSSIP_HEARTBEAT_INTERVAL=] [default: 1]
      --max_transmit_size <MAX_TRANSMIT_SIZE>
          The maximum byte size for each gossip (default is 18 MiB) [env: MAX_TRANSMIT_SIZE=] [default: 18874368]
      --request_timeout <REQUEST_TIMEOUT>
          Choose timeout for sent requests in RequestResponse protocol [env: REQUEST_TIMEOUT=] [default: 20]
      --connection_keep_alive <CONNECTION_KEEP_ALIVE>
          Choose how long RequestResponse protocol connections will live if idle [env: CONNECTION_KEEP_ALIVE=] [default: 20]
      --heartbeat_send_duration <HEARTBEAT_SEND_DURATION>
          Sending of `BlockHeight` should not take longer than this duration, in seconds [env: HEARTBEAT_SEND_DURATION=] [default: 2]
      --heartbeat_idle_duration <HEARTBEAT_IDLE_DURATION>
          Idle time in seconds before sending next `BlockHeight` [env: HEARTBEAT_IDLE_DURATION=] [default: 1]
      --heartbeat_max_failures <HEARTBEAT_MAX_FAILURES>
          Max failures allowed at `Heartbeat` protocol. If reached, the protocol will request disconnect. Cannot be zero [env: HEARTBEAT_MAX_FAILURES=] [default: 5]
      --sync_max_get_header <MAX_GET_HEADER_REQUESTS>
          The maximum number of get header requests to make in a single batch [env: MAX_GET_HEADER_REQUESTS=] [default: 10]
      --sync_max_get_txns <MAX_GET_TXNS_REQUESTS>
          The maximum number of get transaction requests to make in a single batch [env: MAX_GET_TXNS_REQUESTS=] [default: 10]
      --metrics
          [env: METRICS=]
      --verify_max_da_lag <MAX_DA_LAG>
          [env: MAX_DA_LAG=] [default: 10]
      --verify_max_relayer_wait <MAX_WAIT_TIME>
          [env: MAX_WAIT_TIME=] [default: 30s]
      --tx-pool-ttl <TX_POOL_TTL>
          The max time to live of the transaction inside of the `TxPool` [env: TX_POOL_TTL=] [default: 5m]
      --tx-max-number <TX_MAX_NUMBER>
          The max number of transactions that the `TxPool` can simultaneously store [env: TX_MAX_NUMBER=] [default: 4064]
      --tx-max-depth <TX_MAX_DEPTH>
          The max depth of the dependent transactions that supported by the `TxPool` [env: TX_MAX_DEPTH=] [default: 10]
      --tx-number-active-subscriptions <TX_NUMBER_ACTIVE_SUBSCRIPTIONS>
          The maximum number of active subscriptions that supported by the `TxPool` [env: TX_NUMBER_ACTIVE_SUBSCRIPTIONS=] [default: 4064]
      --min_connected_reserved_peers <MIN_CONNECTED_RESERVED_PEERS>
          The number of reserved peers to connect to before starting to sync [env: MIN_CONNECTED_RESERVED_PEERS=] [default: 0]
      --time_until_synced <TIME_UNTIL_SYNCED>
          Time to wait after receiving the latest block before considered to be Synced [env: TIME_UNTIL_SYNCED=] [default: 0s]
      --query_log_threshold_time <QUERY_LOG_THRESHOLD_TIME>
          Time to wait after submitting a query before debug info will be logged about query [env: QUERY_LOG_THRESHOLD_TIME=] [default: 2s]
      --api-request-timeout <API_REQUEST_TIMEOUT>
          Timeout before drop the request [env: API_REQUEST_TIMEOUT=] [default: 30m]
  -h, --help
```

</details>

> _I find it weird that we have no docs about these options on the portal._

Author: Gotcha, added a small note

# Setting up a custom chain

THe `launchNodeAndGetWallets` method lets you launch a local Fuel node with various customizations.

> In the code snippet below, we are providing a custom chain config file to the `launchNodeAndGetWallets` method. You can use a chain config file to customize things like the consensus parameters of the chain or to specify some initial states for the chain. Click here to see what a chain config file looks like: [chainConfig.json](https://github.com/FuelLabs/fuels-ts/blob/master/.fuel-core/configs/chainConfig.json)

Reviewer 3: Tiny nit: active voice instead of passive voice.

```suggestion
In the code snippet below, we provide a custom chain config file to the `launchNodeAndGetWallets` method. You can use a chain config file to customize things like the chain's consensus parameters or specify some initial states for the chain. Click here to see what a chain config file looks like [chainConfig.json](https://github.com/FuelLabs/fuels-ts/blob/master/.fuel-core/configs/chainConfig.json)
```
